Add first draft of NNDM function

mikemahoney218 opened the pull request:
This PR implements a first draft of a function for nearest-neighbor distance matching LOO-CV, as described in Milà et al. 2022 ( https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.13851 ) and implemented in CAST ( https://hannameyer.github.io/CAST/reference/nndm.html ). 

This function produces identical assessment sets to CAST:

```r
data(ames, package = "modeldata")
ames_sf <- sf::st_as_sf(ames, coords = c("Longitude", "Latitude"), crs = 4326)

compare_methods <- function(data, prop) {
  n <- nrow(data)
  train <- sample.int(n, size = floor(n * prop))
  
  purrr::map2_lgl(
    CAST::nndm(data[train, ], ppoints = data[-train, ])$indx_train,
    spatialsample::spatial_nndm_cv(data[train, ], data[-train, ])$splits,
    function(x, y) {
      length(setdiff(x, as.integer(y))) == 0
    }
  ) |> 
    all()
}

all(replicate(25, compare_methods(ames_sf, 0.75)))
#> [1] TRUE
```

My intention here (long-term) is to move the loop into C++, in order to speed up execution time; that said, unless your model is _extremely_ fast, the speed of building resamples is probably not super relevant compared to the speed of fitting models.

mikemahoney218 commented:
@hfrick , any chance you'd be willing to take a look at this one? I always appreciate another set of eyes when adding large new features like this :smile:

hfrick commented on R/spatial_nndm_cv.R:
```diff
@@ -0,0 +1,207 @@
+    # How much data remains in analysis sets?
+    prop_remaining <- sum(!is.na(distance_matrix[indices$row, ])) / n_training
+
+    if ((prop_close_training >= prop_close_prediction) &
```
why is `prop_close_training >= prop_close_prediction` the same as "mean nearest neighbor distance between assessment and analysis folds is still smaller than the mean nearest neighbor distance between training data and the locations a model will be used to predict" (loosely from the documentation (line 5-7))?

mikemahoney218 replied:
Whoops -- it's not the means at all, it's the _distribution_ of nearest neighbor distances. Fixed in the documentation, sorry!

hfrick commented on R/spatial_nndm_cv.R:
```diff
@@ -0,0 +1,207 @@
+
+      nn_training <- apply(distance_matrix, 1, min, na.rm = TRUE)
+
+      indices <- update_indices(indices, nn_training, distance_matrix, `>=`)
```
this corresponds to finding the next-nearest pair, right? That function name is also rather generic and I think it would be easier to read if it was linked more with the domain (= the cross-validation method of nearest neighbor).

hfrick commented on R/spatial_nndm_cv.R:
```diff
@@ -0,0 +1,207 @@
+
+      indices <- update_indices(indices, nn_training, distance_matrix, `>=`)
+    } else {
+      indices <- update_indices(indices, nn_training, distance_matrix, `>`)
```
Can that function have a speaking argument of what the different operators mean? It would be helpful to understand from reading that function call why we update also in the `else` case here.

hfrick reviewed (commented):
Ah, excited that you're adding a new feature like this! 🤩 

I tried to follow along but found this not the most straightforward to read. There are a couple of comments asking for clarification where I was unsure. I think there are a couple of reasons as to why I found this challenging, I'll try to tease those apart:

First and probably most importantly, I did not read that paper so I did not quite know what to expect the code to do when I started reading it. All I knew was what was in the documentation so I tried to match it up with that and with the code comments.

Because of that lack of background knowledge, some early-ish indication of how that selection/elimination process is going to work would have been helpful. There is a lot happening before we get to the core idea of setting a distance entry to `NA` to eliminate an observation from the analysis set. It's not that surprising spelled out like that but I think it would have helped to set the expectation. I'm not sure if that would best happen in a code comment or in a change of names of certain objects and functions. 

The objects and functions handling "the nearest pair" are named so generic that it took me extra effort to link it back to what I (sorta) knew was the point of it all, the nearest neighbor based reduction of the analysis set. Names that reflect the domain more make it easier to place the objects into context and thus make it easier to understand. I left specific comments for those so that I'm not just waffling in the abstract here :D

Let me know if that's helpful!

mikemahoney218 commented:
Thank you so much @hfrick , this is exactly what I was hoping for :smile:

mikemahoney218 commented:
I _think_ I've addressed most of your comments! Let me know if I missed anything. Thank you so much for taking a look at this -- I definitely was bashing my ahead against the wall with this one, and I think some of that in-the-moment confusion unfortunately is reflected in the code :sweat_smile:

hfrick reviewed (approved):
Sure thing! I find that implementing a big new feature requires a lot to be held in my mind at once and that that's not the time I can _also_ think of good names. So I usually do that afterward, once I know stuff's working.
